src/Tuura/Concept/STG/Translation.hs

```diff
@@ -86,9 +86,10 @@ initVal s ls = sum (map (\x -> if (fst x == s) then fromEnum (snd x) else 0) ls)
 readArc :: String -> String -> [String]
 readArc f t = [f ++ " " ++ t, t ++ " " ++ f]
 
-genInvStrs :: (Eq a, Show a) => [[Transition a]] -> [String]
-genInvStrs invs
-        | invs      == [] = []
-        | otherwise = info ++ map (\i -> unwords (map show i)) invs
+genInvStrs :: (Ord a, Show a) => Invariant (Transition a) -> String
+genInvStrs (NeverAll es)
+        | es        == [] = []
+        | otherwise = "# invariant = not (" ++ (format (head sorted)) ++ (concatMap (\e -> " && " ++ format e) (tail sorted)) ++ ")"
     where
-        info = ["Signal transition sets which are invariant:"]
+        format e = if (newValue e) then show (signal e) else "not " ++ show (signal e)
+        sorted = nubOrd es
```

Comment:
Are you using `nubOrd` to simplify the invariant? This looks correct, but I'm not sure we should do that -- let's make this function simpler -- you get what you started with (may be helpful for debugging, for example).